### DISCUSSION TITLE
Configuring nodeSelector globally

### DISCUSSION THREAD

**Discussion by backuitist** [woodpecker-ci/woodpecker#3367] [Ask a Question] [answered]
I have nodes on which no build workload should be scheduled. I couldn't find any way to globally configure pods. I can set the `nodeSelector` in the `backend_options` of every step but beside being very tedious it also doesn't work as some auto-generated steps like `git clone` won't get the proper node selector...

**Comment by zc-devs** [comment]
> globally configure

You can't via Woodpecker now.

> steps like git clone

Doesn't it work if you define [clone step](https://woodpecker-ci.org/docs/usage/workflow-syntax#clone) explicitly?

**Reply by backuitist** [reply]
I guess I could do that, but then every build descriptor must explicitly define the clone step _in addition_ to specifying `backend_options` ...
Setting this globally seems like a reasonable feature?

**Reply by zc-devs** [reply, marked as answer]
> can't via Woodpecker

But you can set up [admission controller](https://stackoverflow.com/questions/52487333/how-to-assign-a-namespace-to-certain-nodes) on [namespace](https://woodpecker-ci.org/docs/administration/backends/kubernetes#woodpecker_backend_k8s_namespace).

**Reply by backuitist** [reply]
Yes, I've done that but couldn't get it to work and it felt a bit hacky (resorting to plugins)... but I could revisit that and create a dedicated namespace for build workload and set it to `WOODPECKER_BACKEND_K8S_NAMESPACE`.
This plus setting the clone step explicitly with a `nodeSelector` (which then forces subsequent steps to run on the same machine), are indeed 2 viable alternatives. Perhaps they could be documented on the k8s backend admin pages. Would you take a PR?
Thanks @zc-devs

**Reply by zc-devs** [reply]
> which then forces subsequent steps to run on the same machine

I suppose, it works like that only in the K3s' Local path provisioner case or something similar. But yes, it could be a hack :)

**Reply by zc-devs** [reply]
> Would you take a PR?

Yes, in general. If it doesn't fit documentation format, you always have `tips & tricks` and `blog post`.

Also be aware of #3376.

**Comment by pat-s** [comment]
Alternatively, you could use https://woodpecker-ci.org/docs/administration/backends/kubernetes#woodpecker_backend_k8s_pod_labels and define some anti-toleration or affinity to prevent WP pods from getting scheduled on some nodes.
